base/linalg/symmetric.jl

```diff
@@ -589,21 +589,53 @@ function ^(A::Hermitian{T}, p::Real) where T
     end
 end
 
-function exp(A::Symmetric)
+for func in (:exp, :cos, :sin, :tan, :cosh, :sinh, :tanh)
+    @eval begin
+        function ($func)(A::Symmetric{T}) where T<:Real
+            F = eigfact(A)
+            return Symmetric((F.vectors * Diagonal(($func).(F.values))) * F.vectors')
+        end
+        function ($func)(A::Hermitian{T}) where T
+            n = checksquare(A)
+            F = eigfact(A)
+            retmat = (F.vectors * Diagonal(($func).(F.values))) * F.vectors'
+            if T <: Real
+                return real(Hermitian(retmat))
+            else
+                for i = 1:n
+                    retmat[i,i] = real(retmat[i,i])
+                end
+                return Hermitian(retmat)
+            end
+        end
+    end
+end
+
+function sincos(A::Symmetric{T}) where T<:Real
+    n = checksquare(A)
     F = eigfact(A)
-    return Symmetric((F.vectors * Diagonal(exp.(F.values))) * F.vectors')
+    S, C = Diagonal(similar(A)), Diagonal(similar(A))
+    for i in 1:n
+        S.diag[i], C.diag[i] = sincos(F.values[i])
+    end
+    return Symmetric((F.vectors * S) * F.vectors'), Symmetric((F.vectors * C) * F.vectors')
 end
-function exp(A::Hermitian{T}) where T
+function sincos(A::Hermitian{T}) where T
     n = checksquare(A)
     F = eigfact(A)
-    retmat = (F.vectors * Diagonal(exp.(F.values))) * F.vectors'
+    S, C = Diagonal(similar(A)), Diagonal(similar(A))
```

Comment:
These could be `similar(A, (n,))`, otherwise they will first allocate a full matrix and then extract the diagonal. Better to simply allocate the diagonal directly.